File: backend/api/outlines.py
from fastapi import APIRouter, Depends

try:
    from api.dependencies import get_active_household, get_system_admin
    from services import outlines
except ModuleNotFoundError:
    from backend.api.dependencies import get_active_household, get_system_admin
    from backend.services import outlines


router = APIRouter()
router.add_api_route(
    "/items/{item_id}/representative-image",
    outlines.get_item_representative_image,
    methods=["GET"],
    dependencies=[Depends(get_active_household)],
)
router.add_api_route(
    "/items/{item_id}/representative-image",
    outlines.upload_item_representative_image,
    methods=["POST"],
    dependencies=[Depends(get_system_admin)],
)
router.add_api_route(
    "/outlines/prepare",
    outlines.start_outline_preparation,
    methods=["POST"],
    dependencies=[Depends(get_system_admin)],
)
router.add_api_route(
    "/outlines/jobs/{job_id}",
    outlines.get_outline_preparation_job,
    methods=["GET"],
    dependencies=[Depends(get_system_admin)],
)

File: mobile/src/services/api/outlines.ts
import type { OutlinePreparationJob } from "../../types/api";
import { requestJson } from "./client";
import { appendUploadFile } from "./upload";

export type { OutlinePreparationJob } from "../../types/api";

let currentOutlinePreparationJobId: string | null = null;

export async function uploadProductRepresentativeImage(
  itemId: number,
  imageUri: string,
) {
  const formData = new FormData();

  await appendUploadFile(
    formData,
    "file",
    imageUri,
    "product-reference.jpg",
    "image/jpeg",
  );

  return requestJson<{ ok: boolean; quality_score: number; outline_revision: string }>(
    `/items/${itemId}/representative-image`,
    {
      method: "POST",
      body: formData,
    },
  );
}

export async function startOutlinePreparation(
  forceNewCheck = false,
): Promise<OutlinePreparationJob> {
  if (!forceNewCheck && currentOutlinePreparationJobId) {
    try {
      return await getOutlinePreparationJob(currentOutlinePreparationJobId);
    } catch {
      currentOutlinePreparationJobId = null;
    }
  }

  const job = await requestJson<OutlinePreparationJob>("/outlines/prepare", {
    method: "POST",
  });

  currentOutlinePreparationJobId = job.job_id;
  return job;
}

export function getOutlinePreparationJob(
  jobId: string,
): Promise<OutlinePreparationJob> {
  return requestJson<OutlinePreparationJob>(`/outlines/jobs/${jobId}`);
}
